routing/index_router.cpp

```diff
@@ -801,14 +760,25 @@ IRouter::ResultCode IndexRouter::RedressRoute(vector<Segment> const & segments,
   return IRouter::NoError;
 }
 
-bool IndexRouter::AreMwmsNear(NumMwmId startId, NumMwmId finishId) const
+bool IndexRouter::AreMwmsNear(std::set<NumMwmId> mwmIds) const
 {
-  m2::RectD const startMwmRect = m_countryRectFn(m_numMwmIds->GetFile(startId).GetName());
-  bool areMwmsNear = false;
-  m_numMwmTree->ForEachInRect(startMwmRect, [&](NumMwmId id) {
-    if (id == finishId)
-      areMwmsNear = true;
-  });
-  return areMwmsNear;
+  // Function fails on first mwm which is not near to all other mwms.
+  // Maximal complexity is O((size of mwms clique)^2).
```

Comment:
From my point of view the comment is wrong, because you're looking in bounding rect, which is an approximation of mwm boundary, so 4 color theorem is not applicable here. Imagine a group of "diagonal" mwms, located near each other, i.e. elleptical mwms with rotated axes.

Reply:
Moreover, even when considering boundaries, our mwms overlap.

Reply:
не знала, теперь буду знать, спасибо